### PR TITLE
SAM reference plant model for HTR-PM reactor

### LINKED ISSUE #428: Add a reference plant model for HTR-PM
## Model Description

This is a reference plant model for the HTR-PM reactor. The model is built with SAM. It contains a porous media multi-D core, a 0-D/1-D primary loop model, and a 0-D/1-D RCCS loop model. The model contains a steady-state simulation and a Pressurized Loss of Forced Cooling (PLOFC) transient simulation. The model runs on a workstation.

## Project description

This model is contributed by Argonne National Lab and is funded by the NRC. This work is continuing into the next FY.

## Impact
This model extends the steady-state and PLOFC simulation beyond the core, which is the commonly used approach for PB-HTGR. The expansion of the model allows a more detailed study on reactor/plant behaviors during various transients.

### PR DESCRIPTION
**Pull request by zhieejhia93**
Closes #428 

This is a SAM reference plant model for the HTR-PM reactor. The model contains a porous media core, a 0D/1D primary loop, and a 0D/1D RCCS loop. A steady-state simulation and a Pressurized Loss of Forced Cooling (PLOFC) are included in this MR.

### CODE REVIEW

**Comment by GiudGiud** [comment]
hello
you ll want to edit the commit message to add "refs #428". this will fix that https://civet.inl.gov/job/2458979/

**Comment by zhieejhia93** [comment]
@GiudGiud I probably misunderstood your comment. I added "refs #428" but still failed precheck

**Comment by GiudGiud** [comment]
different reason this time
```

##########################################################################
ERROR: The following files contain trailing whitespace after applying your patch:
	htgr/htr-pm/ss-main-transient.i
	htgr/htr-pm/ss-primary-loop-transient.i
	htgr/htr-pm/ss-rccs-water-transient.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.

ERROR: The following files do not contain a newline character before EOF:
	htgr/htr-pm/ss-main-transient.i
	htgr/htr-pm/ss-primary-loop-transient.i
	htgr/htr-pm/ss-rccs-water-transient.i
```

File: doc/content/htgr/htr-pm/htr-pm_results.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+!media htrpm_sam/T_multiple_mean.svg
+        style=width:60%
+        id=T_multiple_mean
+        caption=Mean solid structure temperatures during the PLOFC transient.
+
+The temperature increase of the water in the RCCS across the riser is shown in [Trccs_delta]. The temperature profile largely follows the temperature profiles of the solid structures shown in [T_multiple_mean]. As more heat is transferred to the RCCS, the temperature increase rises accordingly until a peak value of roughly 8 K is reached. The RCCS mass flow rate is shown in [mflow_rccs]. As the mass flow rate is driven by the thermal driving head, the profile of the mass flow rate closely follows that of the temperature increase. At its peak, the RCCS has a flow rate of roughly 30 kg/s. The heat removal rate and the decay heat of the reactor are plotted in [heat_rccs]. The heat removal rate exceeds the decay heat at about 75 hours in to the transient, after which the temperatures of the reactor start to come down.
```
Comment (GiudGiud) [inline doc/content/htgr/htr-pm/htr-pm_results.md:88]:
```suggestion
The temperature increase of the water in the RCCS across the riser is shown in [Trccs_delta]. The temperature profile largely follows the temperature profiles of the solid structures shown in [T_multiple_mean]. As more heat is transferred to the RCCS, the temperature increase rises accordingly until a peak increase of roughly 8 K is reached. The RCCS mass flow rate is shown in [mflow_rccs]. As the mass flow rate is driven by the thermal driving head, the profile of the mass flow rate closely follows that of the temperature increase. At its peak, the RCCS has a flow rate of roughly 30 kg/s. The heat removal rate and the decay heat of the reactor are plotted in [heat_rccs]. The heat removal rate exceeds the decay heat at about 75 hours in to the transient, after which the temperatures of the reactor start to come down.
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: doc/content/htgr/htr-pm/htr-pm_model.md [hunk 2]
```diff
@@ -0,0 +1,1687 @@
...
+!media htrpm_sam/DomainOverlap.png
+        style=width:70%
+        id=DomainOverlap
+        caption=Domain overlapping approach used to model fluid flow.
+
+# Problem Description
+
+## Steady-state
+
+A steady-state normal operating condition is simulated with the SAM HTR-PM reference plant model. The operating conditions during steady-state are tabulated in [ss-condition]. The reactor has a power output of 250 MWth and operates at 7 MPa. It is helium cooled with a system mass flow rate of 96 kg/s where the helium enters and leaves the reactor at 523.15 K and 1023.15 K, respectively. The core contains about 420,000 fuel pebbles that are 6 cm in diameter with an average packing fraction of 0.61 [!citep](htrpm_jaradat2023). In this model, neutronics calculations are not powerformed. Instead, a power density distribution obtained from the work by Jaradat et. al. [!citep](htrpm_jaradat2023) is imposed to the 2-D core.
```
[the remark this answers is not in the document]
Comment (zhieejhia93) [reply doc/content/htgr/htr-pm/htr-pm_model.md:53]: I am not sure. There's a generic PBR model by INL but I don't think that is the model that is referenced here.

File: htgr/htr-pm/tests [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    cli_args = '--app SamApp --allow-unused'
+  []
+  [htr_pm_sam_rccs_syntax]
+    type = RunApp
+    input = 'ss-rccs-water.i'
+    check_input = True
+    executable_pattern = 'sam*|blue_crab*'
+    cli_args = '--app SamApp --allow-unused'
+  []
+  [htr_pm_sam]
```
[the remark this answers is not in the document]
Comment (zhieejhia93) [reply htgr/htr-pm/tests]: I added the tests that you asked for.

File: doc/content/htgr/htr-pm/htr-pm_model.md [hunk 4]
```diff
@@ -0,0 +1,491 @@
+# SAM Generic PBR Model
+
+*Contact: Zhiee Jhia Ooi, zooi@anl.gov, Stephen Bajorek, stephen.bajorek@nrc.gov*
+
+*Model link: [HTR-PM SAM Model](https://github.com/idaholab/virtual_test_bed/tree/main/htgr/htr-pm)*
+
+!tag name=SAM HTR-PM Model
+  image=
+  pairs=reactor_type:HTGR
```
Comment (GiudGiud) [inline doc/content/htgr/htr-pm/htr-pm_model.md:9]:
```suggestion
  description=...
  pairs=reactor_type:HTGR
```
forgot about that one

Comment (zhieejhia93) [reply doc/content/htgr/htr-pm/htr-pm_model.md:10]: what should I add in 'image'? I don't see this in other models.

Comment (GiudGiud) [reply doc/content/htgr/htr-pm/htr-pm_model.md:10]: you dont see it locally, but if you rebase you ll see it s in every other model now

Comment (GiudGiud) [reply doc/content/htgr/htr-pm/htr-pm_model.md:10]: you pick a picture, with the path you use for the file in doc/content/media/

Comment (zhieejhia93) [reply doc/content/htgr/htr-pm/htr-pm_model.md:10]: got it. thanks

**Comment by zhieejhia93** [comment]
All comments should be addressed now. Please check again @GiudGiud . Thanks.

File: doc/content/htgr/htr-pm/htr-pm_model.md [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+*Model link: [HTR-PM SAM Model](https://github.com/idaholab/virtual_test_bed/tree/main/htgr/htr-pm)*
+
+!tag name=SAM HTR-PM Model
+  image=https://mooseframework.inl.gov/virtual_test_bed/media/htrpm_sam/loop-schematics-VTB.png
+  description=This is a SAM HTR-PM reference plant model with steady-state and transient PLOFC simulations
+  pairs=reactor_type:HTGR
+                       reactor:HTR_PM
+                       geometry:Core_and_primary_loop
+                       simulation_type:Thermal_hydraulic
```
Comment (GiudGiud) [inline doc/content/htgr/htr-pm/htr-pm_model.md:13]:
```suggestion
                       simulation_type:Thermal_hydraulics
```
or balance of plant?

File: doc/content/htgr/htr-pm/htr-pm_model.md [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+                       transient:PLOFC
```
Comment (GiudGiud) [inline doc/content/htgr/htr-pm/htr-pm_model.md:14]:
```suggestion
                       transient:steady_state;PLOFC
```

File: htgr/htr-pm/tests [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    check_input = True
+    executable_pattern = 'sam*|blue_crab*'
+    cli_args = '--app SamApp --allow-unused'
+  []
+  [htr_pm_sam_main]
+    type = CSVDiff
+    input = 'ss-main.i'
+    csvdiff = 'ss-main_csv.csv'
+    executable_pattern = 'sam*|blue_crab*'
+    cli_args = '--app SamApp --allow-unused Executioner/dt=0.005 Executioner/num_steps=2 Executioner/nl_rel_tol=1e-8 Executioner/nl_abs_tol=1e-7'
```
Comment (GiudGiud) [inline htgr/htr-pm/tests:49]: with those settings, do you have a checlkpoint file you can use?

File: htgr/htr-pm/tests [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    type = CSVDiff
+    input = 'ss-rccs-water.i'
+    csvdiff = 'ss-rccs-water_csv.csv'
+    executable_pattern = 'sam*|blue_crab*'
+    cli_args = '--app SamApp --allow-unused Executioner/dt=0.005 Executioner/num_steps=2 Executioner/nl_rel_tol=1e-8 Executioner/nl_abs_tol=1e-7'
+    abs_zero = 1e-5
+    rel_err = 1e-4
+    max_time = 1000
+    method = 'opt'
+  []
```
Comment (GiudGiud) [inline htgr/htr-pm/tests:76]: we're missing regression tests for the transient input. Can we use an early checkpoint using cli_args to do this?

File: doc/content/htgr/htr-pm/htr-pm_model.md (hunk 5 above)
Comment (zhieejhia93) [reply doc/content/htgr/htr-pm/htr-pm_model.md:13]: ya balance of plant is more appropriate

File: htgr/htr-pm/tests (hunk 8 above)
Comment (zhieejhia93) [reply htgr/htr-pm/tests:76]: Are the checkpoint files stored during testing? If they are stored, the transient tests can be restarted using the checkpoint files generated by the steady-state cases.

Comment (GiudGiud) [reply htgr/htr-pm/tests:76]: they are generated and are not removed.

Comment (GiudGiud) [reply htgr/htr-pm/tests:76]: you can find an example at msr/msre/reactivity_insertion

Comment (GiudGiud) [reply htgr/htr-pm/tests:76]: we use the final checkpoint there iirc but if the steady simulation does not reach it in time, we can set hte number of time steps for the steady state

File: doc/content/htgr/htr-pm/htr-pm_model.md [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+*Contact: Zhiee Jhia Ooi, zooi@anl.gov, Stephen Bajorek, stephen.bajorek@nrc.gov*
+
+*Model link: [HTR-PM SAM Model](https://github.com/idaholab/virtual_test_bed/tree/main/htgr/htr-pm)*
+
+!tag name=SAM HTR-PM Model
+  image=https://mooseframework.inl.gov/virtual_test_bed/media/htrpm_sam/loop-schematics-VTB.png
+  description=This is a SAM HTR-PM reference plant model with steady-state and transient PLOFC simulations
+  pairs=reactor_type:HTGR
+                       reactor:HTR_PM
+                       geometry:Core_and_primary_loop
```
Comment (GiudGiud) [inline doc/content/htgr/htr-pm/htr-pm_model.md:12]: dont make new categories here, you ll want to match one of these
https://mooseframework.inl.gov/virtual_test_bed/resources/filter/index.html

**Comment by zhieejhia93** [comment]
@GiudGiud I reorganized the input files into 'ss' and 'plofc' folders. The 'ss' folder contains the files needed for the steady-state run and the 'plofc' folder files for the PLOFC run. 

Under the 'plofc' folder, in addition to the transient files, there are also steady-state files. This is because a simplified primary loop (instead of the full loop) is used for the PLOFC run. This means that before users can run the PLOFC model, they will have to re-run the steady-state model (using the simplified loop). I thought that separating the files into 'ss' and 'plofc' will be less confusing to users, and will also make testing cleaner.

I also added the syntax and regression tests for all the individual input files.

File: doc/content/htgr/htr-pm/htr-pm_model.md (hunk 9 above)
Comment (zhieejhia93) [reply doc/content/htgr/htr-pm/htr-pm_model.md]: changed to Core;Primary_Loop

File: htgr/htr-pm/tests (hunk 8 above)
Comment (zhieejhia93) [reply htgr/htr-pm/tests]: added individual regression tests for all files

**Comment by GiudGiud** [comment]
I'm ok with the primary loop input being there twice because it's different, but the two other inputs are duplicated.
Either using symbolic links or input file includes (!include syntax) you should be able to avoid that duplication

for testing, I recommend you use a single `tests` file for both plofc and ss, and use the `working_folder` parameter to go to the right folder for each input

**Comment by zhieejhia93** [comment]
I removed the duplicated files and put the remaining files under the same folder. All tests are now in one tests file.

**Review by GiudGiud — APPROVED** [review]
looks great!
Thank you for the contribution!

**Comment by zhieejhia93** [comment]
Thanks for the review!

**Pull request merged**